### PR TITLE
Don't propose disallowed algorithms during hostkey negotiation

### PR DESCRIPTION
**Pull request by ZoltanFridrich**
I would like to propose a patch from rhel-9.1.0

Context:
In rhel-9.1.0 when I try to connect to host that only presents ssh-rsa host keys, the connection fails with a very cryptic error message:
`ssh_dispatch_run_fatal: Connection to 10.19.208.80 port 22: error in libcrypto`
I believe this happens because ssh-rsa is disallowed by crypto-policies but openssh disregards the configuration when negotiating hostkey algorithm. Therefore the ssh-rsa hostkey alg is negotiated and the connection fails when openssl returns an error somewhere inside sshkey_verify function.

This patch restricts the range of negotiated algorithms to only those that are allowed by the configuration. Effectively making the connection fail during hostkey negotiation and with a much more descriptive error message.

Signed-off-by: Zoltan Fridrich <zfridric@redhat.com>

### CODE REVIEW

File: sshconnect2.c [hunk 1]
```diff
@@ -253,16 +253,21 @@ ssh_kex2(struct ssh *ssh, char *host, struct sockaddr *hostaddr, u_short port,
 	    (char *)compression_alg_list(options.compression);
 	myproposal[PROPOSAL_MAC_ALGS_CTOS] =
 	    myproposal[PROPOSAL_MAC_ALGS_STOC] = options.macs;
+
 	if (use_known_hosts_order) {
 		/* Query known_hosts and prefer algorithms that appear there */
-		myproposal[PROPOSAL_SERVER_HOST_KEY_ALGS] =
-		    compat_pkalg_proposal(ssh,
-		    order_hostkeyalgs(host, hostaddr, port, cinfo));
```
Comment (ZoltanFridrich) [inline sshconnect2.c:260]: This patch also fixes a memory leak where allocated pointer returned from order_hostkeyalgs() was being lost without freeing.

Comment (ZoltanFridrich) [reply sshconnect2.c:260]: #324 will need to be resolved first.

File: sshconnect2.c [hunk 2]
```diff
@@ -255,14 +256,19 @@ ssh_kex2(struct ssh *ssh, char *host, struct sockaddr *hostaddr, u_short port,
 	    myproposal[PROPOSAL_MAC_ALGS_STOC] = options.macs;
 	if (use_known_hosts_order) {
 		/* Query known_hosts and prefer algorithms that appear there */
-		myproposal[PROPOSAL_SERVER_HOST_KEY_ALGS] = prop_hostkey =
-		    compat_pkalg_proposal(ssh,
-		    order_hostkeyalgs(host, hostaddr, port, cinfo));
```
Comment (ZoltanFridrich) [inline sshconnect2.c:260]: string returned from order_hostkeyalgs() is lost here

**Comment by ZoltanFridrich** [comment]
The test failures seem expected judging by the output. Basically the negotiation fails because ssh-rsa is not accepted_algo

File: sshconnect2.c [hunk 3]
```diff
@@ -255,14 +256,19 @@ ssh_kex2(struct ssh *ssh, char *host, struct sockaddr *hostaddr, u_short port,
...
+		if ((hostkeyalgs = order_hostkeyalgs(host, hostaddr, port, cinfo)) == NULL)
+			fatal_f("order_hostkeyalgs");
+		pkalg = match_filter_allowlist(hostkeyalgs, options.pubkey_accepted_algos);
```
Comment (djmdjm) [inline sshconnect2.c:261]: This isn't correct: `options.pubkey_accepted_algos` is the `PubkeyAcceptedAlgorithms` configuration option. That is for controlling which signature algorithms are used for user authentication, whereas this is host authentication.

The configuration option that you're looking for is `HostkeyAlgorithms`, but this is already applied inside `order_hostkeyalgs()`

**Comment by djmdjm** [comment]
I added a comment on the change, but this isn't correct. We already have `HostkeyAlgorithms` to control which signature algorithms are used for hostkey negotiation. Your patch instead adds filtering by `PubkeyAcceptedAlgorithms`, which is used for user authentication.

Please just use `HostkeyAlgorithms` for what you're trying to achieve.

**Comment by Jakuje** [comment]
Sorry for being late to the party. To clarify, this issue is related to the previously reported issue https://bugzilla.mindrot.org/show_bug.cgi?id=2924 where we wanted to handle this as part of the `HostkeyAlgorithms` configuration option but it was ignored so this was the second best thing to avoid hard to debug issues, introducing new options.

**Pull request closed without merging**